src/main/haskell/kore/src/Kore/Step/Simplification/Data.hs

```diff
@@ -8,33 +8,69 @@ Stability   : experimental
 Portability : portable
 -}
 module Kore.Step.Simplification.Data
-    ( PureMLPatternSimplifier (..)
+    ( SimplificationError
+    , Simplifier
+    , evalSimplifier
+    , PureMLPatternSimplifier (..)
     , CommonPureMLPatternSimplifier
     , SimplificationProof (..)
     ) where
 
+import Control.Monad.Except
+       ( ExceptT, runExceptT )
+
 import Kore.AST.Common
        ( Variable )
 import Kore.AST.PureML
        ( PureMLPattern )
+import Kore.Error
+       ( Error )
 import Kore.Step.OrOfExpandedPattern
        ( OrOfExpandedPattern )
 import Kore.Variables.Fresh.IntCounter
-       ( IntCounter )
+       ( IntCounter, runIntCounter )
+
+
+{- | A tag for errors during simplification
+
+  See also: 'Error'
+
+ -}
+data SimplificationError
 
 {-| 'SimplificationProof' is a placeholder for proofs showing that the
 simplification of a MetaMLPattern was correct.
 -}
 data SimplificationProof level = SimplificationProof
     deriving (Show, Eq)
 
+{- | The concrete monad in which simplification occurs.
+
+ -}
+-- TODO (thomas.tuegel): Replace IntCounter with a single state carrying both
+-- the counter and the proof.
+-- TODO (thomas.tuegel): Lift the StateT to the outer level.
+type Simplifier = ExceptT (Error SimplificationError) IntCounter
+
+{- | Evaluate a simplifier computation.
+
+  Only the result (or error) is returned. The 'IntCounter' is discarded.
+
+  -}
+evalSimplifier :: Simplifier a -> Either (Error SimplificationError) a
+evalSimplifier simp =
+    let
+        (result, _) = runIntCounter (runExceptT simp) 0
```

Comment:
Could you add a comment saying that we are relying on the IntCounter for generating distinct variable names and, as such, this should not be used between steps or inside step execution? Or something like that.

It would be nice if you could move this in one of the testing modules, but I see that you also used it in Main, so probably a warning is the right thing to do here. Or maybe we could have a special module for things that can only be used in tests and Main...